libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/NecessaryAndSufficient.hs

```diff
@@ -0,0 +1,373 @@
+      getDRepDeposit :: ConwayEraPParams era => PParams era -> Coin
+      getDRepDeposit pp = pp ^. ppDRepDepositL
+   in constrained $ \triple ->
+        match triple $ \env vstate govcert ->
+          match env $ \pp _ ->
+            match vstate $ \voteDelegation _comState _dormantEpochs ->
+              [ satisfies pp pparamsSpec
+              , caseOn
+                  govcert
+                  -- ConwayRegDRep:: Credential 'DRepRole c -> Coin -> StrictMaybe (Anchor c) -> ConwayGovCert c
+                  ( branchW 3 $ \key coin _ ->
+                      [ assertWith 1 status $ not_ $ member_ key (dom_ voteDelegation)
+                      , reify pp getDRepDeposit (\deposit -> assertWith 2 status $ coin ==. deposit)
+                      ]
+                  )
+                  --  ConwayUnRegDRep:: Credential 'DRepRole c -> Coin -> ConayGovCert
+                  ( branchW 6 $ \cred coin ->
+                      [ dependsOn voteDelegation (pair_ cred coin)
+                      , reify voteDelegation getDeposits $
+                          \deposits -> assertWith 3 status $ elem_ (pair_ cred coin) deposits
+                      ]
+                  )
+                  --  ConwayUpdateDRep :: Credential 'DRepRole c -> StrictMaybe (Anchor c) -> ConwayGovCert
+                  ( branchW 2 $ \key _ ->
+                      assertWith 4 status $ member_ key (dom_ voteDelegation)
+                  )
+                  -- ConwayAuthCommitteeHotKey:: Credential 'ColdCommitteeRole c -> Credential 'HotCommitteeRole c -> ConwayGovCert
+                  (branchW 1 $ \c _ -> assert $ (c ==. c))
+                  -- ConwayResignCommitteeColdKey :: Credential 'ColdCommitteeRole c -> StrictMaybe (Anchor c) -> ConwayGovCert
+                  (branchW 1 $ \c _ -> assert $ (c ==. c))
+              ]
+
+govName :: ConwayGovCert c -> String
+govName x = case x of
+  ConwayRegDRep {} -> "ConwayRegDRep"
+  ConwayUnRegDRep {} -> "ConwayUnRegDRep"
+  ConwayUpdateDRep {} -> "ConwayUpdateDRep"
+  ConwayAuthCommitteeHotKey {} -> "ConwayAuthCommitteeHotKey"
+  ConwayResignCommitteeColdKey {} -> "ConwayResignCommitteeColdKey"
+
+-- =================================================================
+-- A test for every Rule
+
+data Status = Apply | Invert Int | Remove deriving (Show)
+
+assertWith :: IsConwayUniv fn => Int -> Status -> Term fn Bool -> Pred fn
+assertWith n Apply x = assertExplain ["Status Apply " ++ show n ++ " (" ++ show x ++ ")"] x
+assertWith n (Invert m) x =
+  if n == m
+    then
+      assertExplain
+        ["Status Invert " ++ show n ++ " (" ++ show x ++ ")"]
+        [assert $ not_ x, monitor (\_eval -> expectFailure)]
+    else
+      assertExplain
+        ["Status Nonmatching Invert (" ++ show n ++ " /= " ++ show m ++ " (" ++ show x ++ ")"]
+        x
+assertWith n Remove x = assertExplain ["Status Remove " ++ show n ++ " (" ++ show x ++ ")"] TruePred
```

Comment:
Wouldn't it make sense to also remove the constraints one-by-one like we do with `Invert`?